File: 01-Numerical_Errors/4-substractive_cancelation.cpp
/* Exploration to the subtractive cancelation in the solutions of ax^2 + bx + c = 0 */
#include<iostream>
#include <cmath>

double x1(double a, double b, double c);
double x2(double a, double b, double c);
double xp1(double a, double b, double c);
double xp2(double a, double b, double c);

int main(int argc, char **argv)
{
	std::cout.precision(15);
	std::cout.setf(std::ios::scientific);
	double a, b, c;
	a = 1;
	b = 1;

	std::cout << "Solutions to the quadratic equation with a=" << a << ", b=" << b << ", c=10^(-n) "<< '\n';
	std::cout << "n" << '\t'
						<< "   	   x_1      " << '\t'
						<< "   	   x_2      " << '\t'
						<< "   	   x'_1     " << '\t'
						<< "   	   x'_2     " << '\n';
	int N = 10;

	for(int n = 1; n<= N; n++)
	{
		c = std::pow(10,-n);


		std::cout << n << '\t'
							<< x1(a,b,c) << '\t'
		 					<< x2(a,b,c) << '\t'
		 					<< xp1(a,b,c) << '\t'
		 					<< xp2(a,b,c) << '\n';
	}

	std::cout << "Most precise solutions: x'_1 & x_2 " << '\n';
	return 0;
}


double x1(double a, double b, double c)
{
	return (-b + std::sqrt(b*b -4*a*c))/(2*a);
}

double x2(double a, double b, double c)
{
	return (-b - std::sqrt(b*b -4*a*c))/(2*a);
}

double xp1(double a, double b, double c)
{
	return -2*c/(b+std::sqrt(b*b - 4*a*c));
}

double xp2(double a, double b, double c)
{
	return -2*c/(b-std::sqrt(b*b - 4*a*c));
}

File: 01-Numerical_Errors/4-substractive_cancelation.py
"""
Description: Exploration to the substactive cancelation in the solutions of ax^2 + bx + c = 0
"""
import numpy as np

def x1(a,b,c):
    return (-b + np.sqrt(b**2 - 4*a*c))/(2*a)

def x2(a,b,c):
    return (-b - np.sqrt(b**2 - 4*a*c))/(2*a)

def xp1(a,b,c):
    return -2*c/(b + np.sqrt(b**2 - 4*a*c))

def xp2(a,b,c):
    return -2*c/(b - np.sqrt(b**2 - 4*a*c))

def main():
    a, b = 1, 1

    N = 10
    print('Solutions to the quadratic equation with a=', a, ', b=', b, ', c = 10^(-n)')
    print('n', '    x_1    ', '     x_2     ', 
          '     x`_1    ', '    x`_2   ' )
    for n in range(1,N+1):
        c = pow(10,-n)
        print(n, "{0:.15f}".format(x1(a,b,c)),
                 "{0:.15f}".format(x2(a,b,c)),
                 "{0:.15f}".format(xp1(a,b,c)),
                 "{0:.15f}".format(xp2(a,b,c)) )
                
main()


"""
//////////////////////////////////////////////////////////////////////////
Created on Fri Mar  14  2021                                    //
                                                                      //
@author: Nicolás Gómez                                               //
//////////////////////////////////////////////////////////////////////
"""
